File: core/src/main/java/org/dreamexposure/discal/core/wrapper/google/EventWrapper.java
package org.dreamexposure.discal.core.wrapper.google;

import com.google.api.client.http.HttpResponse;
import com.google.api.client.http.HttpStatusCodes;
import com.google.api.client.util.DateTime;
import com.google.api.services.calendar.Calendar;
import com.google.api.services.calendar.model.Event;
import org.dreamexposure.discal.core.calendar.CalendarAuth;
import org.dreamexposure.discal.core.logger.LogFeed;
import org.dreamexposure.discal.core.logger.object.LogObject;
import org.dreamexposure.discal.core.object.calendar.CalendarData;
import reactor.core.publisher.Mono;
import reactor.core.scheduler.Schedulers;

import java.util.List;

public class EventWrapper {
    public static Mono<Event> createEvent(CalendarData data, Event event) {
        return CalendarAuth.getCalendarService(data).flatMap(service ->
            Mono.fromCallable(() ->
                service.events()
                    .insert(data.getCalendarId(), event)
                    .setQuotaUser(data.getGuildId().asString())
                    .execute()
            ).subscribeOn(Schedulers.boundedElastic())
        ).doOnError(e -> LogFeed.log(LogObject
            .forException("Event create error; Cred Id: " + data.getCredentialId(), e, EventWrapper.class))
        ).onErrorResume(e -> Mono.empty());
    }

    public static Mono<Event> patchEvent(CalendarData data, Event event) {
        return CalendarAuth.getCalendarService(data).flatMap(service ->
            Mono.fromCallable(() ->
                service.events()
                    .patch(data.getCalendarId(), event.getId(), event)
                    .setQuotaUser(data.getGuildId().asString())
                    .execute()
            ).subscribeOn(Schedulers.boundedElastic())
        ).doOnError(e -> LogFeed.log(LogObject.forException("Failed to patch event", e, EventWrapper.class))
        ).onErrorResume(e -> Mono.empty());
    }

    public static Mono<Event> updateEvent(CalendarData data, Event event) {
        return CalendarAuth.getCalendarService(data).flatMap(service ->
            Mono.fromCallable(() ->
                service.events()
                    .update(data.getCalendarId(), event.getId(), event)
                    .setQuotaUser(data.getGuildId().asString())
                    .execute()
            ).subscribeOn(Schedulers.boundedElastic())
        ).doOnError(e -> LogFeed.log(LogObject.forException("Failed to edit event", e, EventWrapper.class))
        ).onErrorResume(e -> Mono.empty());
    }

    public static Mono<Event> getEvent(CalendarData data, String id) {
        return CalendarAuth.getCalendarService(data).flatMap(service ->
            Mono.fromCallable(() ->
                service.events()
                    .get(data.getCalendarAddress(), id)
                    .setQuotaUser(data.getGuildId().asString())
                    .execute()
            ).subscribeOn(Schedulers.boundedElastic())
        ).onErrorResume(e -> Mono.empty()); //Can ignore this, the event just doesn't exist.
    }

    public static Mono<List<Event>> getEvents(CalendarData data, int amount, long start) {
        return CalendarAuth.getCalendarService(data).flatMap(service ->
            Mono.fromCallable(() ->
                service.events().list(data.getCalendarId())
                    .setMaxResults(amount)
                    .setTimeMin(new DateTime(start))
                    .setOrderBy("startTime")
                    .setSingleEvents(true)
                    .setShowDeleted(false)
                    .setQuotaUser(data.getGuildId().asString())
                    .execute().getItems()
            ).subscribeOn(Schedulers.boundedElastic())
        ).onErrorResume(e -> Mono.empty());
    }

    public static Mono<List<Event>> getEvents(CalendarData data, Calendar service, int amount, long start) {
        return Mono.fromCallable(() ->
            service.events().list(data.getCalendarId())
                .setMaxResults(amount)
                .setTimeMin(new DateTime(start))
                .setOrderBy("startTime")
                .setSingleEvents(true)
                .setShowDeleted(false)
                .setQuotaUser(data.getGuildId().asString())
                .execute().getItems()
        ).subscribeOn(Schedulers.boundedElastic())
            .onErrorResume(e -> Mono.empty());
    }

    public static Mono<List<Event>> getEvents(CalendarData data, int amount, long start, long end) {
        return CalendarAuth.getCalendarService(data).flatMap(service ->
            Mono.fromCallable(() ->
                service.events().list(data.getCalendarId())
                    .setMaxResults(amount)
                    .setTimeMin(new DateTime(start))
                    .setTimeMax(new DateTime(end))
                    .setOrderBy("startTime")
                    .setSingleEvents(true)
                    .setShowDeleted(false)
                    .setQuotaUser(data.getGuildId().asString())
                    .execute().getItems()
            ).subscribeOn(Schedulers.boundedElastic())
        ).onErrorResume(e -> Mono.empty());
    }

    public static Mono<List<Event>> getEvents(CalendarData data, long start, long end) {
        return CalendarAuth.getCalendarService(data).flatMap(service ->
            Mono.fromCallable(() ->
                service.events().list(data.getCalendarId())
                    .setTimeMin(new DateTime(start))
                    .setTimeMax(new DateTime(end))
                    .setOrderBy("startTime")
                    .setSingleEvents(true)
                    .setShowDeleted(false)
                    .setQuotaUser(data.getGuildId().asString())
                    .execute().getItems()
            ).subscribeOn(Schedulers.boundedElastic())
        ).onErrorResume(e -> Mono.empty());
    }

    public static Mono<Boolean> deleteEvent(CalendarData data, String id) {
        return CalendarAuth.getCalendarService(data).flatMap(service ->
            Mono.fromCallable(() -> {
                HttpResponse response = service.events()
                    .delete(data.getCalendarAddress(), id)
                    .setQuotaUser(data.getGuildId().asString())
                    .executeUnparsed();

                //Log error code if one happened
                if (response.getStatusCode() != HttpStatusCodes.STATUS_CODE_OK) {
                    LogFeed.log(LogObject.forDebug(
                            "Event Delete Error | " + response.getStatusCode() + " | " + response.getStatusMessage()
                        ));
                    }

                    return response.getStatusCode() == HttpStatusCodes.STATUS_CODE_OK;
                }
            ).subscribeOn(Schedulers.boundedElastic())
        ).doOnError(e -> LogFeed.log(LogObject.forException("Event Delete Failure", e, EventWrapper.class))
        ).onErrorReturn(false);
    }
}

File: core/src/main/kotlin/org/dreamexposure/discal/core/wrapper/google/EventWrapper.kt
package org.dreamexposure.discal.core.wrapper.google

import com.google.api.client.googleapis.json.GoogleJsonResponseException
import com.google.api.client.util.DateTime
import com.google.api.services.calendar.Calendar
import com.google.api.services.calendar.model.Event
import org.dreamexposure.discal.core.`object`.calendar.CalendarData
import org.dreamexposure.discal.core.logger.LOGGER
import org.dreamexposure.discal.core.utils.GlobalVal
import reactor.core.publisher.Mono
import reactor.core.scheduler.Schedulers
import java.util.function.Predicate

object EventWrapper {
    fun createEvent(calData: CalendarData, event: Event): Mono<Event> {
        return GoogleAuthWrapper.getCalendarService(calData).flatMap { service ->
            Mono.fromCallable {
                service.events()
                        .insert(calData.calendarId, event)
                        .setQuotaUser(calData.guildId.asString())
                        .execute()
            }.subscribeOn(Schedulers.boundedElastic())
        }.doOnError {
            LOGGER.error(GlobalVal.DEFAULT, "[G.Cal] Event create failure", it)
        }.onErrorResume { Mono.empty() }
    }

    fun patchEvent(calData: CalendarData, event: Event): Mono<Event> {
        return GoogleAuthWrapper.getCalendarService(calData).flatMap { service ->
            Mono.fromCallable {
                service.events()
                        .patch(calData.calendarId, event.id, event)
                        .setQuotaUser(calData.guildId.asString())
                        .execute()
            }.subscribeOn(Schedulers.boundedElastic())
        }.doOnError {
            LOGGER.error(GlobalVal.DEFAULT, "[G.Cal] Event patch failure", it)
        }.onErrorResume { Mono.empty() }
    }

    fun updateEvent(calData: CalendarData, event: Event): Mono<Event> {
        return GoogleAuthWrapper.getCalendarService(calData).flatMap { service ->
            Mono.fromCallable {
                service.events()
                        .update(calData.calendarId, event.id, event)
                        .setQuotaUser(calData.guildId.asString())
                        .execute()
            }.subscribeOn(Schedulers.boundedElastic())
        }.doOnError {
            LOGGER.error(GlobalVal.DEFAULT, "[G.Cal] Event update failure", it)
        }.onErrorResume { Mono.empty() }
    }

    fun getEvent(calData: CalendarData, id: String): Mono<Event> {
        return GoogleAuthWrapper.getCalendarService(calData).flatMap { service ->
            Mono.fromCallable {
                service.events()
                        .get(calData.calendarId, id)
                        .setQuotaUser(calData.guildId.asString())
                        .execute()
            }.filter {
                //Don't show "deleted" events
                /*
                See "status" flag: https://developers.google.com/calendar/api/v3/reference/events#resource
                 */
                !it.status.equals("cancelled", true)
            }.subscribeOn(Schedulers.boundedElastic())
        }.onErrorResume { Mono.empty() } //Can safely ignore this, the event just doesn't exist.
    }

    fun getEvents(calData: CalendarData, amount: Int, start: Long): Mono<List<Event>> {
        return GoogleAuthWrapper.getCalendarService(calData).flatMap { service ->
            Mono.fromCallable {
                service.events()
                        .list(calData.calendarId)
                        .setMaxResults(amount)
                        .setTimeMin(DateTime(start))
                        .setOrderBy("startTime")
                        .setSingleEvents(true)
                        .setShowDeleted(false)
                        .setQuotaUser(calData.guildId.asString())
                        .execute()
                        .items
            }.subscribeOn(Schedulers.boundedElastic())
        }.doOnError {
            LOGGER.error(GlobalVal.DEFAULT, "[G.Cal] Event list(1) failure", it)
        }.onErrorResume { Mono.empty() }
    }

    @Deprecated(message = "Deprecated, do not use service directly")
    fun getEvents(calData: CalendarData, service: Calendar, amount: Int, start: Long): Mono<List<Event>> {
        return Mono.fromCallable {
            service.events()
                    .list(calData.calendarId)
                    .setMaxResults(amount)
                    .setTimeMin(DateTime(start))
                    .setOrderBy("startTime")
                    .setSingleEvents(true)
                    .setShowDeleted(false)
                    .setQuotaUser(calData.guildId.asString())
                    .execute().items
        }.subscribeOn(Schedulers.boundedElastic())
                .doOnError {
                    LOGGER.error(GlobalVal.DEFAULT, "[G.Cal] Event list(2) failure", it)
                }.onErrorResume { Mono.empty() }
    }

    fun getEvents(calData: CalendarData, amount: Int, start: Long, end: Long): Mono<List<Event>> {
        return GoogleAuthWrapper.getCalendarService(calData).flatMap { service: Calendar ->
            Mono.fromCallable {
                service.events()
                        .list(calData.calendarId)
                        .setMaxResults(amount)
                        .setTimeMin(DateTime(start))
                        .setTimeMax(DateTime(end))
                        .setOrderBy("startTime")
                        .setSingleEvents(true)
                        .setShowDeleted(false)
                        .setQuotaUser(calData.guildId.asString())
                        .execute().items
            }.subscribeOn(Schedulers.boundedElastic())
        }.doOnError {
            LOGGER.error(GlobalVal.DEFAULT, "[G.Cal] Event list(3) failure", it)
        }.onErrorResume { Mono.empty() }
    }

    fun getEvents(calData: CalendarData, start: Long, end: Long): Mono<List<Event>> {
        return GoogleAuthWrapper.getCalendarService(calData).flatMap { service: Calendar ->
            Mono.fromCallable {
                service.events()
                        .list(calData.calendarId)
                        .setTimeMin(DateTime(start))
                        .setTimeMax(DateTime(end))
                        .setOrderBy("startTime")
                        .setSingleEvents(true)
                        .setShowDeleted(false)
                        .setQuotaUser(calData.guildId.asString())
                        .execute().items
            }.subscribeOn(Schedulers.boundedElastic())
        }.doOnError {
            LOGGER.error(GlobalVal.DEFAULT, "[G.Cal] Event list(4) failure", it)
        }.onErrorResume { Mono.empty() }
    }

    fun deleteEvent(calData: CalendarData, id: String?): Mono<Boolean> {
        return GoogleAuthWrapper.getCalendarService(calData).flatMap { service: Calendar ->
            Mono.fromCallable {
                val response = service.events()
                        .delete(calData.calendarAddress, id)
                        .setQuotaUser(calData.guildId.asString())
                        .executeUnparsed()

                //Google sends 4 possible status codes, 200, 204, 404, 410.
                // First 2 should be treated as successful, and the other 2 as not found.
                when (response.statusCode) {
                    200, 204 -> {
                        return@fromCallable true
                    }
                    404, 410 -> {
                        return@fromCallable false
                    }
                    else -> {
                        //Log response data and return false as google sent an unexpected response code.
                        LOGGER.debug(GlobalVal.DEFAULT, "Event delete error | ${response.statusCode} | ${response.statusMessage}")
                        return@fromCallable false
                    }
                }
            }.subscribeOn(Schedulers.boundedElastic())
        }.doOnError(GoogleJsonResponseException::class.java) {
            if (it.statusCode != 410 || it.statusCode != 404) {
                LOGGER.error(GlobalVal.DEFAULT, "[G.Cal] Event delete failure", it)
            }
        }.doOnError(Predicate.not(GoogleJsonResponseException::class.java::isInstance)) {
            LOGGER.error(GlobalVal.DEFAULT, "[G.Cal] Event delete failure", it)
        }.onErrorReturn(false)
    }
}
